pkg/apis/garden/v1beta1/defaults.go

```diff
@@ -233,3 +253,72 @@ func SetDefaults_VolumeType(obj *VolumeType) {
 		obj.Usable = &trueVar
 	}
 }
+
+// CalculateDefaultNodeCIDRMaskSize calculates a default NodeCIDRMaskSize CIDR from the highest maxPod setting in the shoot
+func CalculateDefaultNodeCIDRMaskSize(kubernetes *Kubernetes, workers []Worker) *int {
+	var maxPod int32
+	if kubernetes.Kubelet != nil && kubernetes.Kubelet.MaxPods != nil {
+		maxPod = *kubernetes.Kubelet.MaxPods
+	}
+
+	for _, worker := range workers {
+		if worker.Kubelet != nil && worker.Kubelet.MaxPods != nil && *worker.Kubelet.MaxPods > maxPod {
+			maxPod = *worker.Kubelet.MaxPods
+		}
+	}
+
+	if maxPod == 0 {
+		// default maxPod setting on kubelet
+		maxPod = 110
+	}
+
+	// by having approximately twice as many available IP addresses as possible Pods, Kubernetes is able to mitigate IP address reuse as Pods are added to and removed from a node.
+	nodeCidrRange := 32 - int(math.Ceil(math.Log2(float64(maxPod*2))))
+	return &nodeCidrRange
+}
+
+// SetNodeCIDRMaskSize sets the NodeCIDRMaskSize on the shoot
+func SetNodeCIDRMaskSize(kubernetes *Kubernetes, requiredNodeCIDRMaskSize *int) {
+	if kubernetes.KubeControllerManager == nil {
+		kubernetes.KubeControllerManager = &KubeControllerManagerConfig{NodeCIDRMaskSize: requiredNodeCIDRMaskSize}
+	} else {
+		kubernetes.KubeControllerManager.NodeCIDRMaskSize = requiredNodeCIDRMaskSize
+	}
+}
+
+// getShootCloudProviderWorkers retrieves the cloud-specific workers of the given Shoot.
+func getShootCloudProviderWorkers(cloudProvider CloudProvider, shoot *Shoot) []Worker {
```

Comment:
Don't reimplement this, use https://github.com/gardener/gardener/blob/master/pkg/apis/garden/v1beta1/helper/helpers.go#L127

Reply:
Problem with cyclic dependencies because the defaults.go is in the same package as the v1beta1 types.

Reply:
I see.. well, then let's live with that. The `garden` API group will be deprecated anyway soon and replaced with a new one where we don't have this issue anymore.